NDTensors/src/UnallocatedArrays/src/unallocated_impl.jl

```diff
@@ -0,0 +1,29 @@
+for Typ in (:UnallocatedFill, :UnallocatedZeros)
+  ## Here are functions specifically defined for UnallocatedArrays
+  ## not implemented by FillArrays
+  @eval begin
+    alloctype(A::$Typ) = alloctype(typeof(A))
+    alloctype(::Type{<:$Typ{ElT,N,Axes,Alloc}}) where {ElT,N,Axes,Alloc} = Alloc
+
+    get_index(A::$Typ) = getindex(data(A))
+
+    array(A::$Typ) = alloctype(typeof(A))(data(A))
```

Comment:
Can't this just be `alloctype(A)`?

Also I don't think it is a good idea to access the wrapped fill array with `data`, `data` has a particular meaning in `NDTensors`. Probably `parent` is better (also same for the name of the field).

Reply:
I wasn't sure if `array` or `Array` should return a type or an initialized version of the Fill. This is what happens in FillArrays
```
julia> f = FillArrays.Fill(2.0, (2,))
julia> Array(f)
  2-element Vector{Float64}:
   2.0
   2.0
```

Reply:
`Array` should definitely return something that satisfies `isa Array`, we should consider that a "law" of calling a constructor.

`array` is slightly underdefined but I would say should also return something that `isa Array`.

Reply:
Are these definitions actually needed in practice?

Reply:
I am not sure they are needed.
I was just thinking about sometimes in `NDTensors` we will call `array(Tensor)` or `matrix(Tensor)` and that the `FillArray's` version has a default value of `Vector` whereas in our `UnallocatedArray` version we store the alloc type so we can instantiate with slightly more robustly, i.e we can directly allocate data in the array type we want instead of making a vector and converting. But we might not want that, I would think using an allocate function would allow us to have more control for where and when data is allocated.